base/special/exp.jl

```diff
@@ -0,0 +1,141 @@
+# Based on FreeBSD lib/msun/src/e_exp.c
+# ====================================================
+# Copyright (C) 2004 by Sun Microsystems, Inc. All rights reserved. Permission
+# to use, copy, modify, and distribute this software is freely granted,
+# provided that this notice is preserved.
+# ====================================================
+
+# Method
+# 1. Argument reduction: Reduce x to an r so that |r| <= 0.5*ln(2). Given x,
+#    find r and integer k such that
+#       x = k*ln(2) + r,  |r| <= 0.5*ln(2).
+#    Here r is represented as r = hi - lo for better accuracy.
+#
+# 2. Approximate exp(r) by a special rational function on [0, 0.5*ln(2)]:
+#       R(r^2) = r*(exp(r)+1)/(exp(r)-1) = 2 + r*r/6 - r^4/360 + ...
+#
+#    A special Remez algorithm on [0, 0.5*ln(2)] is used to generate a
+#    polynomial to approximate R.
+#
+#    The computation of exp(r) thus becomes
+#                       2*r
+#       exp(r) = 1 + ----------
+#                     R(r) - r
+#                          r*c(r)
+#              = 1 + r + ----------- (for better accuracy)
+#                         2 - c(r)
+#    where
+#       c(r) = r - (P1*r^2  + P2*r^4  + ... + P5*r^10 + ...).
+#
+# 3. Scale back: exp(x) = 2^k * exp(r)
+
+# log(2)
+const LN2 = 6.931471805599453094172321214581765680755001343602552541206800094933936219696955e-01
+# log2(e)
+const LOG2E = 1.442695040888963407359924681001892137426646
+
+# log(2) into upper and lower
+LN2U{T<:Float64}(::Type{T}) = T(6.93147180369123816490e-1)
+LN2L{T<:Float64}(::Type{T}) = T(1.90821492927058770002e-10)
+
+LN2U{T<:Float32}(::Type{T}) = T(6.9313812256e-1)
+LN2L{T<:Float32}(::Type{T}) = T(9.0580006145e-6)
+
+# max and min arguments for exponential fucntions
+MAXEXP(::Type{Float64}) = 7.09782712893383996732e2 # log 2^1023*(2-2^-52)
+MAXEXP(::Type{Float32}) = 88.72283905206835f0      # log 2^127 *(2-2^-23)
+
+# one less than the min exponent since we can sqeeze a bit more from the exp function
+MINEXP(::Type{Float64}) = -7.451332191019412076235e2 # log 2^-1075
+MINEXP(::Type{Float32}) = -103.97207708f0            # log 2^-150
+
+# coefficients from: lib/msun/src/e_exp.c
+@inline exp_kernel{T<:Float64}(x::T) = @horner_oftype(x, 1.66666666666666019037e-1,
```

Comment:
Originally my motivation was to have two or more polynomials depending on the float type and use dispatch as follows:
- A low degree polynomial for Float16 and Float32
- A higher degree polynomial for FLoat64 and other Floats like Float128

so in: https://github.com/JuliaMath/Amal.jl/blob/master/src/exp.jl 

I have the two polynomial kernels dispatch the following types
```
SmallFloat = Union{Float16,Float32}
LargeFloat = Float64   # perhaps 
```

That grand goal might be a pipedream. So if you all prefer I can restrict the types instead of using this and the `@horner_oftype` machinery. (although it does have some elegance  to have this  this current machinery with `horner_oftype` and the polynomial dispatch on the type)